src/libraries/Common/src/System/Net/MultiArrayBuffer.cs

```diff
@@ -0,0 +1,319 @@
+        public void TryEnsureAvailableSpaceUpToLimit(int byteCount, int limit)
+        {
+            if (ActiveMemory.Length >= limit)
+            {
+                // Already past limit. Do nothing.
+                return;
+            }
+
+            byteCount = Math.Min(byteCount, limit - ActiveMemory.Length);
+            if (byteCount <= AvailableMemory.Length)
+            {
+                // We have enough space available already.
+                return;
+            }
+
+            int newBytesNeeded = byteCount - AvailableMemory.Length;
+            int newBlocksNeeded = (newBytesNeeded + BlockSize - 1) / BlockSize;
+            Debug.Assert(newBlocksNeeded > 0);
+
+            if (_blocks is null)
+            {
+                Debug.Assert(_blockCount == 0);
+                Debug.Assert(_activeStart == 0);
+                Debug.Assert(_availableStart == 0);
+
+                int blockArraySize = 4;
+                while (blockArraySize < newBlocksNeeded)
+                {
+                    blockArraySize *= 2;
+                }
+
+                _blocks = new byte[]?[blockArraySize];
+            }
+            else if (_blocks.Length < _blockCount + newBlocksNeeded)
+            {
+                int firstUsedBlock = _activeStart / BlockSize;
+                int usedBlockCount = _blockCount - firstUsedBlock;
+                if (usedBlockCount + newBlocksNeeded <= _blocks.Length)
+                {
+                    // We can shift the array down to make enough space
+                    _blocks.AsSpan().Slice(firstUsedBlock, usedBlockCount).CopyTo(_blocks);
+                }
+                else
+                {
+                    // Need to reallocate the array
+                    int blockArraySize = _blocks.Length;
+                    while (blockArraySize < usedBlockCount + newBlocksNeeded)
+                    {
+                        blockArraySize *= 2;
+                    }
+
+                    byte[]?[] newBlockArray = new byte[]?[blockArraySize];
+                    _blocks.AsSpan().Slice(firstUsedBlock, usedBlockCount).CopyTo(newBlockArray);
+                    _blocks = newBlockArray;
+                }
+
+                _blockCount = usedBlockCount;
+                _activeStart -= firstUsedBlock * BlockSize;
+                _availableStart -= firstUsedBlock * BlockSize;
+
+                Debug.Assert(_activeStart / BlockSize == 0, $"Start is not in first block after move or resize?? _activeStart={_activeStart}");
+            }
+
+            Debug.Assert(_blockCount + newBlocksNeeded <= _blocks.Length, $"Not enough room for new blocks?? _blockCount={_blockCount}, newBlocksNeeded={newBlocksNeeded}, _blocks.Length={_blocks.Length}");
+
+            // Allocate new blocks
+            for (int i = 0; i < newBlocksNeeded; i++)
+            {
+                _blocks[_blockCount + i] = ArrayPool<byte>.Shared.Rent(BlockSize);
```

Comment:
This seems a bit aggressive, but I guess it's logically no different than doubling buffer space available, it just happens to be split across N buffers?

Reply:
How is it aggressive? We are allocating as many new blocks as needed to satisfy the user's request for space, no more.

Reply:
Sorry, you're right; I saw a `* 2` earlier and in my head thought that was for number of blocks.